qiskit_experiments/analysis/plotting.py

```diff
@@ -0,0 +1,201 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Plotting functions for experiment analysis
+"""
+import functools
+from typing import Callable, Optional
+import numpy as np
+
+from qiskit_experiments.base_analysis import AnalysisResult
+
+try:
+    from matplotlib import pyplot as plt
+
+    HAS_MATPLOTLIB = True
+except ImportError:
+    HAS_MATPLOTLIB = False
+
+
+def requires_matplotlib(func):
+    """Decorator for functions requiring matplotlib"""
+
+    @functools.wraps(func)
+    def wrapped(*args, **kwargs):
+        if not HAS_MATPLOTLIB:
+            raise ImportError(
+                f"{func} requires matplotlib to generate curve fit plot."
+                ' Run "pip install matplotlib" before.'
+            )
+        return func(*args, **kwargs)
+
+    return wrapped
+
+
+@requires_matplotlib
+def plot_curve_fit(
+    func: Callable,
+    result: AnalysisResult,
+    confidence_interval: bool = True,
+    ax: Optional["AxesSubplot"] = None,
+    num_fit_points: int = 100,
+    labelsize: int = 14,
+    grid: bool = True,
+    **kwargs,
+) -> "AxesSubplot":
+    """Generate plot of a curve fit analysis result.
+
+    Wraps ``matplotlib.pyplot.plot``.
+
+    Args:
+        func: the fit funcion for curve_fit.
+        result: an AnalysisResult from curve_fit.
```

Comment:
Why separate the two? Could we not include the function in the result? It seems like this leaves open the opportunity for having inconsistencies between the result and the function.

Reply:
One reason is the result is a dict that should be serializable and im also not sure how you would store a generic python function in a JSON DB. I think the actual fit function should be defined as part of the Analysis class so I don't think needs to be stored in every result (if you want to you are free to, but i dont think it should be required). I don't think plotting the wrong function is something to worry about too much since these are generally used internally in the Analysis classes.